controllers/idler/aap_idler.go

```diff
@@ -0,0 +1,252 @@
+		if restartCount > AAPRestartThreshold {
+			podLogger.Info("Pod is restarting too often for an AAP pod. Checking if it belongs to AAP and if so then idle the aap", "restart_count", restartCount)
+			aapName, err = r.ensureAAPIdled(podCtx, pod, aapAPI)
+			if err != nil {
+				return err
+			}
+		} else {
+			// Check if running for longer then the AAP idler timeout
+			timeoutSeconds := r.aapTimeoutSeconds(idler)
+			if pod.Status.StartTime != nil && time.Now().After(pod.Status.StartTime.Add(time.Duration(timeoutSeconds)*time.Second)) {
+				podLogger.Info("Pod is running for too long for an AAP pod. Checking if it belongs to AAP and if so then idle the aap.", "start_time", pod.Status.StartTime.Format("2006-01-02T15:04:05Z"), "timeout_seconds", timeoutSeconds)
+				aapName, err = r.ensureAAPIdled(podCtx, pod, aapAPI)
+				if err != nil {
+					return err
+				}
+			}
+		}
+
+		// Check if we need to send a notification and proceed with the next pod
+		if aapName != "" {
+			// The AAP has been Idled
+			// A notification should be sent
+			if err := r.createNotification(podCtx, idler, aapName, "Ansible Automation Platform"); err != nil {
+				podLogger.Error(err, "failed to create Notification")
+				if err = r.setStatusIdlerNotificationCreationFailed(podCtx, idler, err.Error()); err != nil {
+					podLogger.Error(err, "failed to set status IdlerNotificationCreationFailed")
+				} // not returning error to continue tracking remaining pods
+			}
+
+			idledAAPs = append(idledAAPs, aapName)
+			if checkIfAllAAPsIdled(aapList, idledAAPs) {
+				// All AAPs are idled, no need to check the rest of the pods
+				return nil
+			}
+		}
+	}
+
+	return nil
+}
+
+func checkIfAllAAPsIdled(allAAPs *unstructured.UnstructuredList, idledAAPs []string) bool {
+	if len(allAAPs.Items) != len(idledAAPs) {
+		return false
+	}
+	for _, aap := range allAAPs.Items {
+		found := false
+		for _, idledAAP := range idledAAPs {
+			if aap.GetName() == idledAAP {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return false
+		}
+	}
+	return true
+}
+
+const oneHour = 60 * 60 // in seconds
+
+func (r *Reconciler) aapTimeoutSeconds(idler *toolchainv1alpha1.Idler) int32 {
+	// Check if the idler timeout is less than one hour and if so, set it to half of the timeout.
+	// Otherwise, subtract one hour from the timeout.
+	// This is to ensure that the AAP idler kicks in before the main idler.
+	timeoutSeconds := idler.Spec.TimeoutSeconds
+	if timeoutSeconds <= oneHour {
+		timeoutSeconds = timeoutSeconds / 2
+	} else {
+		timeoutSeconds = timeoutSeconds - oneHour
+	}
+	return timeoutSeconds
+}
+
+// ensureAAPIdled checks if the long-running or crash-looping pod belongs to an AAP instance and if so, ensures that the AAP is idled.
+// Returns the AAP resource name in case it was idled, or an empty string if it was not idled.
+func (r *Reconciler) ensureAAPIdled(ctx context.Context, pod corev1.Pod, aapAPI aapAPI) (string, error) {
+	podCondition := pod.Status.Conditions
```

Comment:
we should probably check that the Status and the Conditions are not nil

Reply:
Status can't be nil. It's not a pointer. And range can handle nil. So no need to check for nil here. However I got rid of the podCondition var since we don't really need it: https://github.com/codeready-toolchain/member-operator/pull/631/commits/b1f628febc20e71fabc22ad7495a6541d8f36d74

Reply:
Makes sense! Thanks!